packages/aws-appsync-subscription-link/src/realtime-subscription-handshake-link.ts

```diff
@@ -0,0 +1,745 @@
+        options,
+        observer,
+        subscriptionId
+      });
+
+      return async () => {
+        // Cleanup after unsubscribing or observer.complete was called after _startSubscriptionWithAWSAppSyncRealTime
+        try {
+          this._verifySubscriptionAlreadyStarted(subscriptionId);
+          const { subscriptionState } = this.subscriptionObserverMap.get(
+            subscriptionId
+          );
+          if (subscriptionState === SUBSCRIPTION_STATUS.CONNECTED) {
+            this._sendUnsubscriptionMessage(subscriptionId);
+          } else {
+            throw new Error(
+              "Subscription has failed, starting to remove subscription."
+            );
+          }
+        } catch (err) {
+          this._removeSubscriptionObserver(subscriptionId);
+          return;
+        }
+      };
+    }).filter(data => {
+      const { extensions: { controlMsgType = undefined } = {} } = data;
+      const isControlMsg = typeof controlMsgType !== "undefined";
+
+      return controlEvents === true || !isControlMsg;
+    });
+  }
+
+  private async _verifySubscriptionAlreadyStarted(subscriptionId) {
+    const { subscriptionState } = this.subscriptionObserverMap.get(
+      subscriptionId
+    );
+    // This in case unsubscribe is invoked before sending start subscription message
+    if (subscriptionState === SUBSCRIPTION_STATUS.PENDING) {
+      return new Promise((res, rej) => {
+        const {
+          observer,
+          subscriptionState,
+          variables,
+          query
+        } = this.subscriptionObserverMap.get(subscriptionId);
+        this.subscriptionObserverMap.set(subscriptionId, {
+          observer,
+          subscriptionState,
+          variables,
+          query,
+          subscriptionReadyCallback: res,
+          subscriptionFailedCallback: rej
+        });
+      });
+    }
+  }
+
+  private _sendUnsubscriptionMessage(subscriptionId) {
+    try {
+      if (
+        this.awsRealTimeSocket &&
+        this.awsRealTimeSocket.readyState === WebSocket.OPEN &&
+        this.socketStatus === SOCKET_STATUS.READY
+      ) {
+        // Preparing unsubscribe message to stop receiving messages for that subscription
+        const unsubscribeMessage = {
+          id: subscriptionId,
+          type: MESSAGE_TYPES.GQL_STOP
+        };
+        const stringToAWSRealTime = JSON.stringify(unsubscribeMessage);
+        this.awsRealTimeSocket.send(stringToAWSRealTime);
+
+        this._removeSubscriptionObserver(subscriptionId);
+      }
+    } catch (err) {
+      // If GQL_STOP is not sent because of disconnection issue, then there is nothing the client can do
+      logger({ err });
+    }
+  }
+
+  private _removeSubscriptionObserver(subscriptionId) {
+    this.subscriptionObserverMap.delete(subscriptionId);
+    if (this.subscriptionObserverMap.size === 0) {
+      // Socket could be sending data to unsubscribe so is required to wait until is flushed
+      this._closeSocketWhenFlushed();
```

Comment:
But in this waiting period, another subscription could happen in this socket, then it will close the new subscription. 😢